### PR TITLE
Fix field accessor for RVA field

### LINKED ISSUE #103207: FieldAccessor can't read RVA field correctly
https://github.com/dotnet/runtime/pull/102739#discussion_r1617545620

```csharp
var array = new int[] { 1, 2, 3, 4, 5 }; // Uses RVA field for initialization

foreach (var type in typeof(Program).Assembly.GetTypes())
{
    if (type.Name.Contains("PrivateImplementationDetails"))
    {
        foreach (var field in type.GetFields(System.Reflection.BindingFlags.Static | System.Reflection.BindingFlags.NonPublic))
        {
            _ = field.GetValue(null);
            _ = field.GetValue(null);
        }
    }
}
```

Currently, the first call is correctly handled by the slow path. Subsequent calls are handled by FieldAccessor and failing. The problems are:
- `RuntimeFieldHandle.GetStaticFieldAddress` is returning the RVA offset, not its address. Trying to access it will cause access violation, and translated to `NullReferenceException` because it's typically at page 0.
- The managed side of `FieldAccessor` tries to read the field as boxed. RVA fields are not boxed.

Test is added in #102739. I'm already working on the fix.

### PR DESCRIPTION
**Pull request by huoyaoyuan**
Fixes #103207

### CODE REVIEW

File: src/libraries/System.Runtime/tests/System.Reflection.Tests/FieldInfoTests.cs [hunk 1]
```diff
@@ -121,6 +123,29 @@ public void GetValueWithFunctionPointers(Type type, string name, object obj, obj
             Assert.Equal(expected, fieldInfo.GetValue(obj));
         }
 
+        [Fact]
+        public void GetValueFromRvaField()
+        {
+            byte[] valueArray = new byte[] { 1, 2, 3, 4, 5 };
+
+            // Roslyn uses SHA256 of raw data as data field name
```
Comment (steveharter) [inline src/libraries/System.Runtime/tests/System.Reflection.Tests/FieldInfoTests.cs:131]: Interesting; this is likely why the original RVA tests added with the fast field feature didn't fail.

**Review by steveharter — APPROVED** [review]
Thanks!

**Review by steveharter — CHANGES_REQUESTED** [review]
There is a test failure on macOS-12.7.2. Does Mono do things differently?
```
[FAIL] System.Reflection.Tests.FieldInfoTests.GetValueFromRvaField
System.NotSupportedException : This object cannot be invoked because no code was generated for it: '<PrivateImplementationDetails>.74F81FE167D99B4CB41D6D0CCDA82278CAEE9F3E2F25D5E5A3936FF3DCEC60D0'.
   at System.Reflection.Runtime.FieldInfos.RuntimeFieldInfo.get_FieldAccessor() + 0xe8
   at System.Reflection.Runtime.FieldInfos.RuntimeFieldInfo.GetValue(Object) + 0x14
   at System.Reflection.Tests.FieldInfoTests.GetValueFromRvaField() + 0x120
   at System.Reflection!<BaseAddress>+0xb29b40
   at System.Reflection.DynamicInvokeInfo.Invoke(Object, IntPtr, Object[], BinderBundle, Boolean) + 0x10c
```

**Comment by huoyaoyuan** [comment]
It's NativeAOT. Seems that NativeAOT doesn't support reflection on RVA field. Let's see if it can easily be added or just disable the test.

@dotnet/ilc-contrib

**Comment by jkotas** [comment]
> Seems that NativeAOT doesn't support reflection on RVA field. Let's see if it can easily be added or just disable the test.

I do not think that it is something we want enable. Could you please disable the test for NAOT?

**Comment by MichalStrehovsky** [comment]
> It's NativeAOT. Seems that NativeAOT doesn't support reflection on RVA field. Let's see if it can easily be added or just disable the test.

Agreed with Jan. This would only be useful for two things: C++/CLI support and support for (non-intrinsic) RuntimeHelpers.InitializeArray. Native AOT has both listed as unplanned in https://github.com/dotnet/runtime/issues/69919.

The whole reflection with RVA static field business would ideally be just blocked everywhere, it has always been buggy - calling SetValue on RVA statics on .NET Framework crashes the runtime badly, and accessing RVA statics in the TLS range doesn't work as expected (accessing fields marked `thread_local` in C++/CLI will read the template default value instead of the current value of the threadstatic).

**Comment by huoyaoyuan** [comment]
The failures look unrelated now.

**Review by jkotas — APPROVED** [review]

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Review by steveharter — APPROVED** [review]

**Pull request merged**